src/chain/shared_state.rs

```diff
@@ -377,23 +429,48 @@ impl SectionProofChain {
         }
         true
     }
-}
 
-impl Debug for SectionProofChain {
-    fn fmt(&self, formatter: &mut Formatter) -> fmt::Result {
-        write!(
-            formatter,
-            "SectionProofChain(len = {})",
-            self.blocks.len() + 1
-        )
+    pub fn last_public_key(&self) -> &BlsPublicKey {
+        self.blocks
+            .last()
+            .map(|block| &block.key)
+            .unwrap_or(&self.genesis_pk)
+    }
+
+    pub fn last_index(&self) -> usize {
+        self.blocks.len()
+    }
+
+    pub fn all_keys(&self) -> impl DoubleEndedIterator<Item = &BlsPublicKey> {
+        iter::once(&self.genesis_pk).chain(self.blocks.iter().map(|block| &block.key))
+    }
+
+    pub fn len(&self) -> usize {
+        self.blocks.len() + 1
+    }
+
+    pub fn slice(&self, first_index: usize, last_index: usize) -> SectionProofChain {
+        let genesis_pk = if first_index == 0 {
+            self.genesis_pk.clone()
+        } else {
+            self.blocks[first_index - 1].key.clone()
+        };
+        let blocks = if last_index == 0 {
+            vec![]
+        } else if last_index < self.blocks.len() {
+            self.blocks[first_index..last_index].to_vec()
```

Comment:
don't get it why need to check against `last_index` ? I thought 
`
let blocks = if last_index == 0 {
    vec![]
} else {
    self.blocks[first_index..].to_vec()
};
`
will be enough?

Reply:
Because of the way we use it, we could pass only first_index the the slice function and call it differently.
But as implemented slice needs to be able to return a slice that does not include the end.

Reply:
Ah, I realized the block I attached was wrong, it shall be:
`
let blocks = if last_index == 0 { vec![] } else { self.blocks[first_index..last_index].to_vec() };
`
however, this requires the `last_index` to be the real index that includes itself, not `the end` that excludes itself.